hnn_core/extracellular.py

```diff
@@ -0,0 +1,551 @@
+            # the section axis, enforce minimal perpendicular distance
+            R2 = np.maximum(R2, min_distance ** 2)
+
+            if L < 0 and H < 0:  # electrode is "behind" line segment
+                num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+                denom = np.sqrt(L ** 2 + R2) - L
+            elif L > 0 and H < 0:  # electrode is "on top of" line segment
+                num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+                denom = R2
+            else:  # electrode is "ahead of" line segment
+                num = np.sqrt(L ** 2 + R2) + L
+                denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+            phi[idx] = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class ExtracellularArray:
+    """Class for recording extracellular potential fields with electrode array
+
+    Note that to add an electrode array to a simulation, you should use the
+    :meth:`hnn_core.Network.add_electrode_array`-method. After simulation,
+    the network will contain a dictionary of `ExtracellularArray`-objects
+    in ``net.rec_array`` (each array must be added with a unique name). An
+    `ExtracellularArray` contains the voltages at each electrode contact,
+    along with the time points at which the voltages were sampled.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the extracellular electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. ``'psa'`` (point source approximation) treats
+        each segment junction as a point extracellular current source.
+        ``'lsa'`` (line source approximation) treats each segment as a line
+        source of current, which extends from the previous to the next segment
+        center point: |---x---|, where x is the current segment flanked by |.
+    min_distance : float (default: 0.5; unit: um)
+        To avoid numerical errors in calculating potentials, apply a minimum
+        distance limit between the electrode contacts and the active neuronal
+        membrane elements that act as sources of current. The default value of
+        0.5 um corresponds to 1 um diameter dendrites.
+    times : None | list of float
+        Optionally, provide precomputed voltage sampling times for electrodes
+        at `positions`.
+    voltages : None | list of list of list of float (3D)
+        Optionally, provide precomputed voltages for electrodes at `positions`.
```

Comment:
I didn't quite get what's the usecase for this

Reply:
This is my fault, I missed the `get_data` method at my first read through. 

Personally I find it more natural to access voltage data through an attribute, as the method really doesn't add anything to this basic functionality. If `get_data`did some sort of processing (smoothing, downsampling, etc.) I think it'd be justified, but as it stands it seems like unnecessary lines of code.

Reply:
Actually, the `times` and `voltages` _constructor arguments_ are necessary to enable slicing using `__getitem`.

The `voltages` _property_ returns the contents of `self._voltages`, which currently is a list. The `get_data` _method_ returns the voltages as a Numpy array. The optional argument `return_times` can be used to also return the times as a Numpy array.